profiling/src/profiling/stack_walking.rs

```diff
@@ -88,6 +88,42 @@ pub fn extract_function_name(func: &zend_function) -> Option<Cow<'static, str>>
     }
 }
 
+/// Safely get opline reference with bounds checking to prevent segfaults on dangling pointers that
+/// have been observed when dereferencing `execute_data.opline` under some conditions.
+///
+/// Safety: Relies on the caller to only try and access an `opline` for a non internal function
+#[inline]
+fn safely_get_opline(execute_data: &zend_execute_data) -> Option<&crate::bindings::zend_op> {
+    if execute_data.opline.is_null() {
+        return None;
+    }
+
+    let func = unsafe { execute_data.func.as_ref()? };
+
+    // Safety: relies on the caller to only call for non internal functions. In that case, the
+    // `op_array` is used in the union.
+    let op_array = unsafe { &func.op_array };
+    let opcodes_start = op_array.opcodes;
+
+    if opcodes_start.is_null() {
+        return None;
+    }
+
+    // Safety: `opcodes_start` is null checked and both point/pointed to the same allocated object.
+    let opline_offset = unsafe { execute_data.opline.offset_from(opcodes_start) };
```

Comment:
Is this true though? Isn't that why we have issues in the first place, because the opline doesn't point inside there?

Reply:
The main problem is that we don't guarantee the alignment here.
I'd propose
```suggestion
    // Safety: `opcodes_start` is null checked and both point/pointed to the same allocated object.
    let opline_offset = unsafe { execute_data.opline.byte_offset_from(opcodes_start) / size_of::<T>() };
```
to avoid the casting. That function has no such requirements.

Reply:
Yes, it was pointing into it, but it might not anymore and then the alignment is not guaranteed. Thanks

Reply:
If I read this right, I don't think `byte_offset_from` is any better:

> Calculates the distance between two pointers within the same allocation. The returned value is in units of bytes.

They aren't in the same allocation. I think you need to do `(self as isize - origin as isize) / size_of::<T>()` as [offset_from](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset_from-1) recommends.

Reply:
We might not need to calculate an offset after all, just check the pointers:
```rust
    let opcodes_end = unsafe { opcodes_start.add(op_array.last as usize) };
    if execute_data.opline >= opcodes_start && execute_data.opline < opcodes_end {
```
This looks more like [PHP is doing it](https://heap.space/xref/PHP-8.0/ext/opcache/jit/zend_jit_trace.c?r=072dc3c857a00d687ba00a62e73d5051294bcdfb#7141-7142):
```c
ZEND_ASSERT(opline >= EX(func)->op_array.opcodes &&
	opline < EX(func)->op_array.opcodes + EX(func)->op_array.last);
```

Reply:
In C, I believe it's implementation defined behavior to do this comparison. I'm not _sure_ what Rust says about it, but I think it's UB. I know you can use a pointer that is one-past the end of the end of the allocated object's range, and can do comparisons and arithmetic with it, but that doesn't mean you can do it with arbitrary pointers, like we're doing here.